src/Network/Transport/TCP.hs

```diff
@@ -1753,6 +1765,26 @@ withScheduledAction ourEndPoint f =
           (traverse (\(tp, a) -> runScheduledAction (ourEndPoint, tp) a) <=< readIORef)
           (\ref -> f (\rp g -> mask_ $ schedule rp g >>= \x -> writeIORef ref (Just (rp,x)) ))
 
+-- | Run an IO but kill it after an optional timeout in microseconds.
+withTimeout :: Exception e => e -> Int -> IO t -> IO t
+withTimeout e timeout io = mask $ \unmask -> do
+  var <- newEmptyMVar
+  rec {
+      tidTimeout <- forkIO $ do
+        unmask (threadDelay timeout >> killThread tidAction)
+        putMVar var Nothing
+    ; tidAction  <- forkIO $ do
+        -- If there's an exception, it'll be thrown up top when we takeMVar
+        -- and pattern-match on the value within.
+        outcome <- unmask $
+          (io `catch` (throw :: SomeException -> IO t)) <* killThread tidTimeout
```

Comment:
`killThread` should probably not run with exception unmasked, unless `withTimeout` has as precondition that `io` won't die of asynchronous exceptions. Also, `killThread` might be interruptible, so it should be masked uninterruptibly.

`withTimeout` will also leave the spawned threads running if interrupted. The threads should be cleaned up, or it should be added as a precondition that `withTimeout` should not be interrupted.

None of this affects the current use of `withTimeout`.